### DISCUSSION TITLE
"Error running query Oracle"  does not show the line with error

### DISCUSSION THREAD

**Discussion by lscapim** [getredash/redash#6367] [General]
Hi,
When we run a query in Redash, with errors, the error message does not show the line, as in the example:
<img width="251" alt="image" src="https://github.com/getredash/redash/assets/96078362/cd5cda36-3ea5-4c98-b5d7-66b89abada92">
<img width="228" alt="image" src="https://github.com/getredash/redash/assets/96078362/b9298067-21ae-4d2e-9ca6-6eb59d4a0d9a">

Can we find a why to solve this? 
When we have a big query this can be a big problem for us.

Thank you

**Comment by justinclift** [comment]
@snickerjp As our resident Oracle specialist, what do you think? :smile:

**Comment by snickerjp** [comment]
@lscapim 
This report does not show the "location of the error" in the error message,
You don't know where the error occurred.

"I would like you to show me the location of the error."
Is that correct?

Certainly, I agree with you.

**Comment by snickerjp** [comment]
How can I make it so that I can output the same error messages as in SQL*Plus?
I have looked into this.

Redash uses the Python library "cx_Oracle" to connect to Oracle Database.

It is not possible to output the same error message as sqlplus using cx_Oracle. The error message from cx_Oracle contains the Oracle Database error number and error message. On the other hand, the error message from sqlplus contains the message corresponding to the Oracle Database error code.

The Python module "python-oracledb" was the same.

If there is a better way, please let me know.

I hope this helps.

The following was used as a reference
https://cx-oracle.readthedocs.io/en/latest/api_manual/module.html#exceptions

**Comment by snickerjp** [comment]
In the course of my investigation, I made some modifications to the Python code.
I was able to display **the Offset number**.

~~~diff
diff --git a/redash/query_runner/oracle.py b/redash/query_runner/oracle.py
--- a/redash/query_runner/oracle.py
+++ b/redash/query_runner/oracle.py
@@ -176,7 +176,8 @@ class Oracle(BaseSQLQueryRunner):
                 json_data = json_dumps(data)
                 connection.commit()
         except cx_Oracle.DatabaseError as err:
-            error = "Query failed. {}.".format(str(err))
+            err_args, = err.args
+            error = "Query failed. {}.".format(str(err)) + "| Error Offset num: {}.".format(str(err_args.offset))
             json_data = None
         except (KeyboardInterrupt, JobTimeoutException):
             connection.cancel()
~~~

However, perhaps this is not enough.

![image](https://github.com/getredash/redash/assets/1247622/15e7330f-d344-4f3c-8b15-bb08526ff263)

**Comment by justinclift** [comment]
With that offset number, I'm guessing that's the number of (unicode?) characters into the query text?

Although it's not exactly at the helpfulness of something like "line 52, column 10", it still sounds like an improvement over the existing situation.

Did you want to make a PR for that change? :smile:

**Comment by lscapim** [comment]
@justinclift It's exactly the same situation you mentioned. We need to have the line and column returned with error, similar to the example that showed and SQLPlus: "line 52, column 10"

**Comment by justinclift** [comment]
@snickerjp With that offset number that it returns... that might actually be useful "enough" for our purposes if we're clever.

@eradman If a data source tells us which character # in our query is the error point, do you reckon we could (accurately) work our the line and column numbers from that?  I'm thinking it's *probably* not too hard to write a function for calculating that (front end?) if such a function doesn't already exist.

Thoughts?

**Reply by eradman** [reply]
Is this problem unique to Redash? I assume Oracle returning an offset number instead of a line number.

**Reply by justinclift** [reply]
Not sure.  I'd imagine it wouldn't be unique to us, as it seems to be information being provided back to us via the Oracle data source library.

**Reply by snickerjp** [reply]
I think this case depends on the Oracle data source library (cx_Oracle).

I don't think this is a Redash unique problem.

**Reply by snickerjp** [reply]
To display the offset number
I will try to create an pull request.

**Reply by snickerjp** [reply]
Created pull request. #6373

**Reply by lscapim** [reply]
@snickerjp  Thank you very much... if we can somehow display the line where the script error is it will help a lot. In oracle we can see the line with the error.
<img width="256" alt="image" src="https://github.com/getredash/redash/assets/96078362/b91fa44e-00db-44a2-aad0-4f16ea77fcb4">

**Comment by justinclift** [comment]
Thanks to @snickerjp for creating the PR to add the line and column location in the error message output (#6373).  It's been merged now.

Next, we need to figure out the release process for creating Redash builds people can use in production. :smile:
